src/Neo/SmartContract/Native/GasToken.cs

```diff
@@ -43,6 +43,14 @@ internal override async ContractTask OnPersist(ApplicationEngine engine)
             {
                 await Burn(engine, tx.Sender, tx.SystemFee + tx.NetworkFee);
                 totalNetworkFee += tx.NetworkFee;
+
+                // Reward for NotaryAssisted attribute will be minted to designated notary nodes
+                // by Notary contract.
+                var notaryAssisted = tx.GetAttribute<NotaryAssisted>();
+                if (notaryAssisted is not null)
+                {
+                    totalNetworkFee -= (notaryAssisted.NKeys + 1) * Policy.GetAttributeFee(engine.Snapshot, (byte)notaryAssisted.Type);
```

Comment:
it could underflow, isn't it?

Reply:
We guarantee that `NetworkFee` pays for `(nkeys + 1) * FEE` when checking the NetworkFee value (see `CalculateNetworkFee` for the attribute).

Reply:
Imagine that the networkFee is 1000 with one NKey, and the transaction is persisted in the second position ob the block and the Attribute fee is 100, it will burn 900, it's ok.

Now imagine that in the first position of the block, the commitee change the fee of this attribute to 10_000.
1000 - 10_000 = -9000 <- underflow

Reply:
Yeah, I had this thought about fee value change, but this seems to be a more systemic problem in that if we use fees as they're defined by the current state including previously processed transactions then transactions can fail because of invalid fee (or we can mint excessive GAS to Oracle price change). For other cases we always considered these as too rare to care. But you're right in that it's more dangerous here and should be handled even if we leave other cases as is. The question though, what's the best way? Just going zero? And then capping the overall reward amount to the network fee in notary contract? I don't see any other valid options unless we can get the old fee value.

Reply:
@shargon, @roman-khimov, but how it's possible if all calculations are made in `OnPersist` (for both Gas and Notary contracts). So we capture the NotaryAssisted attribute price in OnPersist (*not in PostPersist*), i.e. before any in-block transactions processing. So I don't see the underflow problem here since any subsequent in-block transactions won't affect the GAS distribution.

Reply:
True. We check `N+1` block transactions relative to the state after `N` execution. Then `OnPersist` runs _before_ transactions, so committee can't change the fee (only some other native contract technically can, but none of them does so) and this game can't be played. Even GAS distribution in the Notary contract is in `OnPersist` as well (unlike Oracle's, btw).

Reply:
Can we add a unit test for this?

Reply:
Yes, will add, but it's better to be added in #3178 since there's a fully-qualified Notary contract.

Reply:
This was the fee to the notary nodes, isn't it? why don't force to transfer the gas first and avoid this change?

Reply:
You mean transfer it right here? The idea was to leave fee _distribution_ to the Notary contract (the scheme is rather simple, but the less code we have in GAS contract the better). So in GAS we just deal with this attribute, other attributes can have some similar rules and also do redistribution wherever it's more appropriate. We can change it, there will just be more code in the GAS contract.

Reply:
I discussed this with @vang1ong7ang as well, maybe this can be handled outside the  GAS Contract.

Reply:
It's a network fee and network fees are handled in the GAS contract. I don't see any way to avoid it. Network fee makes payment guaranteed and makes sense in general (we're using some service provided by the network here). System fee is specific to execution.